internal/controllers/gateway/gateway_utils.go

```diff
@@ -398,6 +454,42 @@ func getListenerStatus(
 	return statusArray
 }
 
+// getReferenceGrantConditionReason gets a certRef belonging to a specific listener and a slice of referenceGrants.
+func getReferenceGrantConditionReason(gatewayNamespace string, certRef gatewayv1alpha2.SecretObjectReference, referenceGrants []gatewayv1alpha2.ReferenceGrant) string {
+	// no need to have this reference granted
+	if certRef.Namespace == nil || *certRef.Namespace == gatewayv1alpha2.Namespace(gatewayNamespace) {
+		return string(gatewayv1alpha2.ListenerReasonResolvedRefs)
+	}
+
+	certRefNamespace := string(*certRef.Namespace)
+	for _, grant := range referenceGrants {
+		// the grant must exist in the same namespace of the referenced resource
+		if grant.Namespace != certRefNamespace {
+			continue
+		}
+		for _, from := range grant.Spec.From {
+			// we are interested only in grants for gateways that want to reference secrets
+			if from.Group != gatewayV1alpha2Group || from.Kind != "Gateway" {
+				continue
+			}
+			if from.Namespace == gatewayv1alpha2.Namespace(gatewayNamespace) {
+				for _, to := range grant.Spec.To {
+					if (to.Group != "" && to.Group != "core") || to.Kind != "Secret" {
+						continue
+					}
+					// if all the above conditions are satisfied, and the name of the referenced secret matches
+					// the granted resource name, then return a reason "ResolvedRefs"
+					if to.Name == nil || string(*to.Name) == string(certRef.Name) {
+						return string(gatewayv1alpha2.ListenerReasonResolvedRefs)
+					}
+				}
+			}
+		}
+	}
+	// if no grants have been found for the reference, return an "InvalidCertificateRef" reason
+	return string(gatewayv1alpha2.ListenerReasonInvalidCertificateRef)
```

Comment:
It doesn't seem that we added any unit tests for this function, I recommend we add some.